### PR TITLE
fix(timesheet): honor backdate restriction boundary on frontend, with holiday/leave-aware backend computation

### LINKED ISSUE #883: Visually disable backdated timesheet entries beyond limit
# 🚀 Feature Request

## **Description**
In the next-pms timesheet view, the frontend should visually disable (gray out/make read-only) the dates that fall beyond the permitted backdated time entry limit. 

## **Motivation**
Currently, there is no visual indication on the UI to show which past dates are locked for time entry. Users can spend time manually filling out their timesheets for past dates, only to receive a "Backdated timeentry not allowed" error upon submission. Disabling the inputs directly on the UI will prevent this wasted effort, reduce frustration, and significantly improve the overall user experience.

## **Mockups or References**

Image

## **Acceptance Criteria:**
- [x] The frontend fetches the backdated time entry limit configuration from the backend.
- [x] Timesheet columns/cells for dates older than the allowed limit are visually indicated as disabled (e.g., grayed out).
- [x] Users are prevented from clicking, adding, or editing time entries on these disabled dates (inputs should be strictly read-only or disabled).
- [ ] (Optional) Provide a tooltip when hovering over disabled dates explaining why they are locked (e.g., "Backdated time entry limit exceeded").

## **Additional Context**
The backend already correctly validates this by throwing a `Backdated timeentry not allowed` error. This feature request is specifically to handle this validation gracefully on the frontend (FE) before the user even attempts to input data.

### PR DESCRIPTION
**Pull request by PraveenKum11**
### Backend
- Adds `get_backdate_restriction_boundary(employee)` as the single source of truth for the
  earliest date a user may still log time against - reused by `validate_dates` (enforcement)
  and exposed via `get_data` / team / project APIs (frontend precheck).
- Boundary computation counts working days only, skipping holidays, weekly-offs, and
  approved/open leave, matching backend validation exactly.
- Fixes the Ignored Role / Administrator exemption not being honored by the boundary
  calculation - exempt users were still seeing dates disabled in the grid despite the
  backend accepting the entry.
- Adds test coverage: boundary math (employee/manager thresholds, master toggle, zero-day
  edge cases), exemptions, holiday/leave widening, and `validate_dates` integration.

### Frontend
- Threads `backdate_restricted_before` down through the timesheet grid's row components
  (task, project, member, week, totals) so dates before the boundary render disabled.
- Uses a single shared `isDateBackdateRestricted` check against the backend-computed boundary.

### But why did i do **Backend-computed boundary instead of client-side accumulation**?

I first tried making the frontend fetch each employee's holidays and leave, then re-count "working days" itself in JS to
match the backend's rule. I dropped this because it meant the same logic lived in two places
(Python and JS), and they could quietly fall out of sync over time + a lot of complexity was added to achieve this.

Instead, the backend now does this counting once and just tells the frontend a single date: "you
can't add entries before this date." This works because the restriction only ever gets stricter
the further back you go, never looser - so one date is all the frontend needs, not the full
holiday/leave calendar.

This way, the holiday/leave logic lives in exactly one place (the backend), the frontend just
compares a date to that one boundary, and any future backend fix (like the Ignored Role bug)
automatically works correctly on the frontend too, with no frontend changes needed.

## Screenshot/Screencast

https://github.com/user-attachments/assets/c5306fd5-59f3-4ecb-8c91-7c88735c6669

## Checklist

- [x] I have carefully reviewed the code before submitting it for review.
- [x] This code is adequately covered by unit tests to validate its functionality.
- [x] I have conducted thorough testing to ensure it functions as intended.
- [ ] A member of the QA team has reviewed and tested this PR (To be checked by QA or code reviewer)

Fixes #883

### CODE REVIEW

File: next_pms/timesheet/doc_events/timesheet.py [hunk 1]
```diff
@@ -142,45 +131,106 @@ def validate_dates(doc):
...
+
+    working_days_counted = 0
+    current_date = today_date
+    while working_days_counted < allowed_days:
+        current_date = add_days(current_date, -1)
+        if current_date < search_start:
+            break
+        if str(current_date) not in non_working_days:
+            working_days_counted += 1
+    return str(current_date)
```
[the remark this answers is not in the document]
Comment (PraveenKum11) [reply next_pms/timesheet/doc_events/timesheet.py]: fixed.

File: next_pms/timesheet/api/team.py [hunk 2]
```diff
@@ -217,6 +221,7 @@ def build_team_member_payload(employee, context, week, has_filters):
         "tasks": detail.get("tasks", {}),
         "leaves": list(context["leaves_by_employee"].get(employee.name, [])),
         "holidays": list(context["holidays_by_employee"].get(employee.name, [])),
+        "backdate_restricted_before": get_backdate_restriction_boundary(employee.name),
```
[the remark this answers is not in the document]
Comment (PraveenKum11) [reply next_pms/timesheet/api/team.py]: fixed.

File: next_pms/timesheet/api/project.py [hunk 3]
```diff
@@ -313,6 +314,7 @@ def _build_project_employee_payload(
         "working_hours": working_hours,
         "holidays": list(context["holidays_by_employee"].get(employee.name, [])),
         "leaves": list(context["leaves_by_employee"].get(employee.name, [])),
+        "backdate_restricted_before": get_backdate_restriction_boundary(employee.name),
```
[the remark this answers is not in the document]
Comment (PraveenKum11) [reply next_pms/timesheet/api/project.py]: fixed.

**Review by b1ink0 — APPROVED** [review]
LGTM FE changes

**Review by wreckage0907 — APPROVED** [review]
